### PR TITLE
feat: Added cron specific endpoint

### LINKED ISSUE #2101: [Crons] Implement a monitor checkin API endpoint
The goal of this is to implement a convenience API endpoint in relay for users to create monitor check-ins **without the complexities of constructing an envelope** or having to explicitly use a SDK.

### Goals

Implement a bare-bones endpoint that does the following:

### 1. Receives the minimum necessary information to generate a check-in

- `sentry_public_key`
- [`monitor_slug`](https://github.com/getsentry/relay/blob/faa1de98f6ecfa6e62cd88248913b76747f0f911/relay-monitors/src/lib.rs#L111)
- [`status`](https://github.com/getsentry/relay/blob/faa1de98f6ecfa6e62cd88248913b76747f0f911/relay-monitors/src/lib.rs#L44-L53)

### Endpoint design details

We would like to keep the URL as short as possible, since these URLs will typically end up in users shell scripts, it's ideal for it to be short and clear.

#### URL design

For the initial pass we will **only support `GET` requests**.

We will support 2 ways to authenticate check-ins

```bash
# Public key is part of the --user argument. This is desirable as it will
# keep the public key out of logs. While the key is public, it may still be
# ideal to not continually log it.
#
# Note the leading `:`! if this is left off curl will prompt for a password.
# Can also be trailing if that is simpler to implement
curl -u :5833ef4ec765499aa7c06b32a8a8879a https://oXXX.ingest.sentry.io/api/cron/{monitor_slug}/?status={status}

# For users who are not concerned about logging their key, it may be
# provided in the URL as a parameter
curl https://oXXX.ingest.sentry.io/api/cron/{monitor_slug}:5833ef4ec765499aa7c06b32a8a8879a/?status={status}
```

**Note:** When we initially discussed this, we discussed using the monitor GUID as an identifier. Because we'll be using the public key (DSN) we will have all the information we need to route the monitor check-in using it's user configured `slug` instead. This is preferred as it makes it much easier to identify at a glance which monitor this endpoint is generating check-ins for.

**Note:** For this first pass, let's just implement both types of endpoints and play with it internally a little bit and see how it feels. We can decide to keep both, remove one, or add a different style if things don't ergonomically feel right.

#### Project ID

If the project ID **absolutely must be kept** like other endpoints it should be placed after `/api/{project_id}/cron/...`. This does lengthen the URL however, so ideall this is NOT required.

#### Extra credit

To shorten the URL even more, it could be very nice to support reading the public key as a base62 token. In our examples above the token would shorten like so:

```bash
# base16: 5833ef4ec765499aa7c06b32a8a8879a
# base62: 2gR3exZ1xSi16PZIDS3cnS

curl https://oXXX.ingest.sentry.io/api/cron/{monitor_slug}:2gR3exZ1xSi16PZIDS3cnS/?status={status}

# An example
curl https://oXXX.ingest.sentry.io/api/cron/customer-invoices:2gR3exZ1xSi16PZIDS3cnS/?status=ok
```

This is not absolutely required for the initial implementation, but if it's simple enough to implement across relay for consistency, it would be nice.

#### Query parameters

* `status`
  This parameter is required and MUST be one of the valid check-in status. Otherwise 400

* `environment`
  Not required, we will default to `production` when it is not set.

#### Response

`202 Accepted` - On success. No body

`400 Bad Request` - On invalid request (status is bad, invalid public key, if we're able to detect that?). The response should be a plain textt

### 2. Constructs the appropriate monitor envelope and ingests it for processing within relay.

<details>
<summary>See the `relay-monitors` crate</summary>

https://github.com/getsentry/relay/blob/faa1de98f6ecfa6e62cd88248913b76747f0f911/relay-monitors/src/lib.rs#L105-L127

</details>

### Check-in ID

The client SDK typically handles opening / closing monitor check-ins by generating a check-in UUID. We do not want to do that here. Instead we will support a similar approach to the existing legacy endpoints, where a sentinel value `latest` can be provided as the `check_in_id`.

This is not yet supported in the monitor consumer (https://github.com/getsentry/sentry/issues/49664). This **will be required to be completed** before we can fully end to end test this endpoint in relay.

## Future work

As mentioned earlier, we should keep in mind that in the near future we would like to expand this endpoint. Since we would like to deprecate the old endpoints in sentry, we will need to support all the features they do.

Likely we will support `POST` requests with a JSON body to allow for:

 * Monitor upserts
 * Check-in ID specification
 * Generating checkins and responding on the API

Since we will just be initially implementing a GET, this should leave things pretty open ended.

### LINKED ISSUE #49203: Crons: Relay Updates Phase 2 - New Ingest Endpoint

### PR DESCRIPTION
**Pull request by mitsuhiko**
Implements a cron specific API endpoint. The `check_in_id` is changed to nullable in this. If a `check_in_id` is not supplied, the consumer will get the nil `check_in_id` which is an alias for "latest". I decided not to tackle the base62 implementation, mostly because now that this reuses the generic infrastructure it might be quite confusing if we end up with different key formats.

The URL format is also slightly different than in the original ticket. The following options are supported:

```
curl -u sentry_key: https://oXXX.ingest.sentry.io/api/cron/{monitor_slug}/?status={status}
curl https://oXXX.ingest.sentry.io/api/cron/{monitor_slug}/?status={status}&sentry_key={sentry_key}
curl https://oXXX.ingest.sentry.io/api/cron/{monitor_slug}/{sentry_key}/?status={status}
```

Work remaining:

* [x] tests
* [x] `latest` for `check_in_id`
* [x] requires split up of some monitors code for non `processing` builds

Fixes #2101
Fixes https://github.com/getsentry/sentry/issues/49203

### CODE REVIEW

File: relay-server/src/extractors/request_meta.rs
Comment (evanpurkhiser) [inline relay-server/src/extractors/request_meta.rs:496]: It looks like we still have the encoding stuff in here. Did I misunderstand that you wanted to skip implementing that?

Comment (evanpurkhiser) [reply relay-server/src/extractors/request_meta.rs:496]: nvm this is from the header for basic auth 👍

**Comment by evanpurkhiser** [comment]
Looking over this I think we're good to merge this.

I'll create some follow up tickets since not long after we'll likely also want to support `POST`ing to this endpoint to support all the features we have in the existing endpoints (passing monitor configurations for upserts)

File: relay-server/src/extractors/request_meta.rs
Comment (mitsuhiko) [reply relay-server/src/extractors/request_meta.rs:496]: @evanpurkhiser this is because HTTP requires `foo:bar` to be sent base64 encoded: https://en.wikipedia.org/wiki/Basic_access_authentication

**Comment by mitsuhiko** [comment]
@evanpurkhiser fwiw I did implement some parameters as query strings. Eg as implemented you can pass `?check_in_id=....` for instance. (See all in `CronQuery`). But I do agree that a post endpoint would be nice.

File: CHANGELOG.md
Comment (olksdr) [inline CHANGELOG.md:9]: There is the unnecessary new line before this record

File: relay-server/src/endpoints/mod.rs
Comment (olksdr) [inline relay-server/src/endpoints/mod.rs:62]: Why do we have to support trailing `/` and no trailing `/` URLs?  Have the routes without trailing slashes been  already exposed to customers and customer got used to using them by now? 

If not,  IMO we have to keep our endpoints consistent, and steer the customers to use only one defined route. And currently we have `/` at the end, which I would also lean to.

File: relay-server/src/extractors/request_meta.rs
Comment (olksdr) [inline relay-server/src/extractors/request_meta.rs:484]: What the reason even to add basic auth? 

Using basic auth is super confusion to me to be honest. 

Especially when you look at the format users should use to make it work. And I talk about the part when you have yo prove the key (`curl -u sentry_key: `) this colon after `sentry_key` is super easy to miss and mess up your curl command. IMO it makes this very error-prone to the customer. 

Much easier and more explicit way would be to use the header which we support or path param, which is also supported.

File: relay-server/src/endpoints/cron.rs
Comment (olksdr) [inline relay-server/src/endpoints/cron.rs:74]: Do we want to have this comment here?

**Review by olksdr — CHANGES_REQUESTED** [review]
Requesting changes, since I have still few questions I would like to clarify before merging this into the Relay.

File: relay-server/src/endpoints/mod.rs
Comment (jernejstrasner) [reply relay-server/src/endpoints/mod.rs:62]: I generally think, from a devex perspective, it's confusing to get errors for missing (or extra) slashes at the end of an URL. Although, I would have imagined we handle this in a cleaner way that adding both URL variants.

Comment (evanpurkhiser) [reply relay-server/src/endpoints/mod.rs:62]: Agree would be nice to handle this in a more general way

File: relay-server/src/extractors/request_meta.rs
Comment (evanpurkhiser) [reply relay-server/src/extractors/request_meta.rs:484]: We're adding this as a bit of an experiment for crons to feel out the ergonomics and DX doing it this way.

I don't think we're planning to immediately tell customers to use this. I agree with the points you make though @olksdr, I think it's not unlikely this will get removed.

**Comment by evanpurkhiser** [comment]
@olksdr Anything else you wanted to follow up with on this? @mitsuhiko I think we're good to go ahead and merge this in and start playing with it yeah?

File: relay-server/src/extractors/request_meta.rs
Comment (olksdr) [reply relay-server/src/extractors/request_meta.rs:484]: Please, add the note here, which explains why this added and that this is just an experiment and will be removed in the follow-up.

File: relay-server/src/endpoints/cron.rs
Comment (olksdr) [reply relay-server/src/endpoints/cron.rs:74]: @evanpurkhiser can we remove this comment? And return what you want to return? Or add a note here to followup and why the return value cannot be decided right now.

**Review by olksdr — APPROVED** [review]
Left few comments. 
Will approve to unblock you, if the open questions will be followed up on. 

And also maybe @jan-auer wants also have a quick review

**Review by olksdr — APPROVED** [review]
Left few comments. 
Will approve to unblock you, if the open questions will be followed up on. 

And also maybe @jan-auer wants also have a quick review

File: tests/integration/test_crons.py
Comment (evanpurkhiser) [inline tests/integration/test_crons.py:47]: Just noticed, this should be `get` right @mitsuhiko ?

File: relay-server/src/extractors/request_meta.rs
Comment (mitsuhiko) [reply relay-server/src/extractors/request_meta.rs:484]: @olksdr fwiw I think basic auth should be generally supported, beyond this experiment, entirely independent of this. From the original `x-sentry-auth` header basically nothing remains and we already support auth purely by `?sentry_key` in the URL. At this point I would honestly love to see us supporting just `Bearer` and basic auth with just the public key.

cc @jan-auer

File: tests/integration/test_crons.py
Comment (mitsuhiko) [reply tests/integration/test_crons.py:47]: The endpoint supports either.

Comment (evanpurkhiser) [reply tests/integration/test_crons.py:47]: Ah cool. I'll expand on POST in a future PR.

**Comment by evanpurkhiser** [comment]
Will follow up on your comments @olksdr Thanks!

**Pull request merged**
